tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,340 @@
+      )
+      .call(this.yAxis);
+    this.xAxisGroup
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left},
+      ${this.drawMargin.top + this.chartHeight})`
+      )
+      .call(this.xAxis);
+    this.miscGroup
+      .append('text')
+      .style('fill', 'black')
+      .text('NaN')
+      .attr('font-size', '10px')
+      .attr('text-anchor', 'end')
+      .attr('alignment-baseline', 'middle')
+      .attr('x', -5)
+      .attr('y', this.chartHeight - this.drawMargin.top);
+    this.mainContainer
+      .append('line')
+      .style('stroke', 'grey')
+      .style('stroke-dasharray', '3, 3')
+      .attr('x1', 0)
+      .attr('y1', this.drawHeight + this.drawMargin.top)
+      .attr('x2', this.chartWidth)
+      .attr('y2', this.drawHeight + this.drawMargin.top);
+  }
+
+  drawPlot() {
+    this.dotsGroup.selectAll('*').remove();
+    this.dotsGroup
+      .selectAll('.violin-plot')
+      .data(Object.keys(this.chartData.violinData))
+      .enter()
+      .append('g')
+      .attr('class', 'violin-plot')
+      .attr(
+        'transform',
+        function(this: ViolinFilterComponent, d: any) {
+          return 'translate(' + this.xScale(d) + ' ,0)';
+        }.bind(this)
+      )
+      .append('path')
+      .style(
+        'stroke',
+        function(this: ViolinFilterComponent, d: any) {
+          return `rgba(${this.rgbColors[0]}, 1.0)`;
+        }.bind(this)
+      )
+      .style(
+        'fill',
+        function(this: ViolinFilterComponent, d: any) {
+          return `rgba(${this.rgbColors[0]}, 0.3)`;
+        }.bind(this)
+      )
+      .datum(
+        function(this: ViolinFilterComponent, d: any) {
+          return this.chartData.violinData[d];
+        }.bind(this)
+      )
+      .attr(
+        'd',
+        d3
+          .area()
+          .x0(
+            function(this: ViolinFilterComponent, d: any) {
+              return this.xScaleNum(-d.length);
+            }.bind(this)
+          )
+          .x1(
+            function(this: ViolinFilterComponent, d: any) {
+              return this.xScaleNum(d.length);
+            }.bind(this)
+          )
+          .y(
+            function(this: ViolinFilterComponent, d: any) {
+              if (d.x0 === -Infinity) {
+                return this.chartHeight - this.drawMargin.top;
+              }
+              return this.yScale((d.x1 + d.x0) / 2.0);
+            }.bind(this)
+          )
+          .curve(d3.curveCatmullRom)
+      );
+  }
+
+  drawBrush() {
+    let brush = d3
+      .brushY()
+      .extent([
+        [
+          this.margin.left + this.drawMargin.left,
+          this.margin.top + this.drawMargin.top,
+        ],
+        [
+          this.width - this.margin.right,
+          this.height - this.margin.bottom / 2.0,
+        ],
+      ])
+      .on('end', this.updateBrush.bind(this));
+    this.svg.selectAll('.brush').remove();
+    this.svg
+      .append('g')
+      .attr('class', 'brush')
```

Comment:
this method that gets called every time in redraw is needlessly removing and reinserting DOM. Can we instead `select`, `enter`, `exit`, and `forEach` to avoid such needless mutations?